### DISCUSSION TITLE
can I spawn an existing machine as an actor?

### DISCUSSION THREAD

**Discussion by csenio** [statelyai/xstate#3067] [Q&A]
- Every UI component in my app has its own machine. 
- Some of these machines need to be able to communicate

Now what I did was have 1 big machine that contains all the other machines and can receive and send events to the machines that need them.

Unfortunately, this is a bit ugly in my app. Since I need to `spawn` an actor for each ui component, it means that on the very first render, my component won't have access to its own machine.

So I have 
1. component is rendered
2. component sends "rendered" event to the parent machine
3. parent machine spawns an actor for the component
4. actor is passed back to the component
5. component can now work 

Is there a simpler way I can coordinate multiple UI components with their own machines? 
Or is there a way I can use my machine on the first render and pass the existing machine to the parent machine?

(I'm using svelte BTW, but it should be the same in React or vue)

**Comment by davidkpiano** [comment]
In my opinion, instead of trying to keep two parallel "trees" (the component tree and the actor system tree), start a machine for the component and use _actions_ to communicate with the parent. This is easiest if there is no need for the parent to communicate back with the child machine.

It gets a bit trickier the other way (and sending events from parent -> child -> parent is tricky, regardless of whether you use XState or not). Can you share more info or example code for your use-case?

**Reply by csenio** [reply]
Sure, here's 1 scenario that this is needed for:

I have a `select` component inside of a `modal dialog` component. 
- Given both are open
- When a user presses `ESC` 
- then only the `select` should close

![sketch](https://user-images.githubusercontent.com/43379421/154850682-6cdf4d4a-80a2-46ef-90a1-ecb8e68e8a6c.jpeg)

Here are some more scenarios:

- Given A `select` is open inside a `modal dialog`
- When the `modal dialog` is closed
- the `select` component should also close

Or one that doesn't involve the modal:
- Given you have multiple `select` components on a page
- When a user opens a `select` component
- All other `select` components should close


It is possible to solve these differently, mainly by listening to all user on the page in each machine, but I think having one machine to coordinate my components makes things much easier to read and more robust.

**Reply by csenio** [reply]
@davidkpiano I have another use case with the same issue:
For this example I have a bunch of audio clips. 
Each audio clip can either be `loading` `idle` `playing` or `paused`. 
What makes this hard is that
- only 1 clip can be playing at once
- once 1 clip finishes, the next one automatically starts

![audio (1)](https://user-images.githubusercontent.com/43379421/156687860-fd775484-cccb-4fc0-9fc9-ec90c3e23a6f.jpeg)

I could do this with just one big machine and use `context` to store the state of each audio clip, but this again seems like a perfect application for the actor model.

**Reply by SimeonC** [reply]
I created a little experiment out of curiosity. https://codepen.io/SimeonC/pen/OJzyqpP??editors=0010

You "can" kind of workaround doing this like you want (I think). The workaround is in essence manually setting the `id` and `parent` values of the child actor. This is needed as when `spawn` is called with a running machine the parent and id won't be set. These only get set if you spawn a machine config. https://github.com/statelyai/xstate/blob/b36ef9dda560fca4c00428f48742fd9d2e325324/packages/core/src/interpreter.ts#L1287-L1291

**Reply by Andarist** [reply]
@jescowuester Did you know that you can spawn machines inside a lazy initial context? I think this could help your use case. This can be achieved using a code like this:
```js
createMachine({
  context: () => {
    return { actorRef: spawn(someMachine) }
  }
})
```
This way you shouldn't have a problem accessing this `actorRef` during your initial render as this isn't spawned in response to an event.

> You "can" kind of workaround doing this like you want (I think). The workaround is in essence manually setting the id and parent values of the child actor

This might work (although I wouldn't recommend it, since it's abusing some private APIs) to some extent but at this point - what's the reason to "bind" a child to a parent? Just to make `sendParent` work? Or are there any other potential benefits of such a thing?

**Reply by SimeonC** [reply]
Yea, now I think about it it was just to get `sendParent` to work so it could be done via `withConfig({ actions: { someAction: () => parentService.send(...) } })`. I did realize I could do this without "hacks" by using the `InterpreterOptions` argument and passing "parent" there. Updated example https://codepen.io/SimeonC/pen/NWXxLga

Is using `meta._event.origin` considered abusing private APIs? 🤔 
Though my example is "vanilla", in React land I could see this being used more like the following, or even the "coordinator" machine being interpreted outside react instead of a context. I get that we can workaround this by doing `withConfig` and manually creating send actions like above, but I think this way is easier to understand and work with IMO.
I think that being able to rely on the session id we don't have to worry about manually created ids and passing those around everywhere. I feel like being able to do this kind of use case leads to simpler to read code as all we read is `sendParent` - though you could also argue that the logic get's a bit more difficult to follow as we need to keep track of "which" machine is the parent.
```tsx
const Component = () => {
  const coordinator = useContext(CoordinatorServiceContext);
  const [state, send] = useMachine(playerMachine, { parent: coordinator });
  ... rest of component rendering
}
```

**Reply by Andarist** [reply]
> Yea, now I think about it it was just to get sendParent to work so it could be done via withConfig({ actions: { someAction: () => parentService.send(...) } }).

Yeah, and this solution also comes with an improved typing experience. It's because you actually can only send the events that this `parentService` accepts and not **any** event.

In general, I'm trying to rethink how we can improve the safety of cross-actor communication and APIs like `sendParent` don't make that any easier. We might have to consider deprecating it in the future in favor of more explicit APIs that would play better with TS.

> I did realize I could do this without "hacks" by using the InterpreterOptions argument and passing "parent" there

I know that this ain't annotated as private or anything - but I would consider this to be rather an internal option. This won't suddenly stop working but it is slightly abusing internal APIs. We need this for our own purposes at the moment but it was never meant for public consumption.

> Is using meta._event.origin considered abusing private APIs? 🤔

This is a public API - but it also won't give you any type-safety at the moment.

> I think that being able to rely on the session id we don't have to worry about manually created ids and passing those around everywhere. I feel like being able to do this kind of use case leads to simpler to read code as all we read is sendParent - though you could also argue that the logic get's a bit more difficult to follow as we need to keep track of "which" machine is the parent.

I agree. Things like `sendParent` are quite convenient - but at the expense of being not quite type-safe. I need to rethink this more thoroughly because we need to balance type safety with good developer experience and this is quite tricky here.
